price-feeder/oracle/oracle.go

```diff
@@ -33,15 +56,256 @@ func (o *Oracle) Stop() {
 	<-o.closer.Done()
 }
 
-func (o *Oracle) Start(ctx context.Context) {
+// SetPrices retrieve all the prices from
+// our set of providers as determined in the config,
+// average them out, and update the oracle object
+func (o *Oracle) SetPrices() error {
+	providerPrices := make(map[string]map[string]sdk.Dec)
+	wg := new(sync.WaitGroup)
+
+	for providerName, tickerList := range o.providerPairs {
+		wg.Add(1)
+		var priceProvider provider.Provider
+
+		switch providerName {
+		case "binance":
+			priceProvider = provider.NewBinanceProvider()
+		case "kraken":
+			priceProvider = provider.NewKrakenProvider()
+		}
+
+		prices, err := priceProvider.GetTickerPrices(tickerList...)
+		if err != nil {
+			return err
+		}
+
+		providerPrices[providerName] = prices
+		wg.Done()
```

Comment:
I like your approach with the waitgroup! But a few things:

- wait groups are useful when used with goroutines (threads) -- we're not using any here.
- we should use [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) instead since `GetTickerPrices` returns an error.
- So ultimately, this would look something like:

```go
func (o *Oracle) SetPrices() error {
	g := new(errgroup.Group)
	mtx := new(sync.Mutex)
	providerPrices := make(map[string]map[string]sdk.Dec)

	for providerName, currencyPairs := range o.providerPairs {
		var priceProvider provider.Provider

		switch providerName {
		case config.ProviderBinance:
			priceProvider = provider.NewBinanceProvider()

		case config.ProviderKraken:
			priceProvider = provider.NewKrakenProvider()
		}

		g.Go(func() error {
			ticker := make([]string, len(currencyPairs))
			for _, cp := range currencyPairs {
				ticker = append(ticker, cp.String())
			}

			prices, err := priceProvider.GetTickerPrices(ticker...)
			if err != nil {
				return err
			}

			// flatten and collect prices based on the base currency per provider
			//
			// e.g.: {ProviderKraken: {"ATOM": 34.03, "UMEE": 6.3}}
			mtx.Lock()
			for _, cp := range currencyPairs {
				if p, ok := prices[cp.String()]; ok {
					providerPrices[providerName][cp.Base] = p
				}
			}
			mtx.Unlock()

			return nil
		})
	}

	if err := g.Wait(); err != nil {
		return err
	}

	// consolidate the different provider maps into one for each exchange rate
	var (
		priceAverages = make(map[string]sdk.Dec)
		priceCounts   = make(map[string]int)
	)

	// TODO: Consider using Volume Weighted Average Price (VWAP).
	//
	// Ref: https://github.com/umee-network/umee/issues/251
	for _, prices := range providerPrices {
		for base, price := range prices {
			if _, ok := priceAverages[base]; !ok {
				priceAverages[base] = sdk.NewDec(0)
			}

			priceAverages[base] = priceAverages[base].Add(price)
			priceCounts[base]++
		}
	}

	for k, average := range priceAverages {
		average = average.QuoInt64(int64(priceCounts[k]))
	}

	o.prices = priceAverages

	return nil
}
```

Assume we have the following:

```go
type CurrencyPair struct {
  Base string
  Quote string
}

func (cp CurrencyPair) String() string {
  return cp.Base + cp.Quote
}
```